File: src/main/java/com/ra4king/circuitsim/gui/CircuitSimVersion.java
package com.ra4king.circuitsim.gui;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

public final class CircuitSimVersion implements Comparable<CircuitSimVersion> {
	private static final Pattern VERSION_PATTERN = Pattern.compile("(\\d+)\\.(\\d+)\\.(\\d+)(b?)");
	
	public static final CircuitSimVersion VERSION = new CircuitSimVersion("1.10.0-CE");

	private final String version;
	private final int major;
	private final int minor;
	private final int bugfix;
	private final boolean beta;
	
	public CircuitSimVersion(String version) {
		this.version = version;
		
		Matcher matcher = VERSION_PATTERN.matcher(version);
		
		if (!matcher.find()) {
			throw new IllegalArgumentException("Invalid version string: " + version);
		}
		
		major = Integer.parseInt(matcher.group(1));
		minor = Integer.parseInt(matcher.group(2));
		bugfix = Integer.parseInt(matcher.group(3));
		beta = !matcher.group(4).isEmpty();
	}
	
	public String getVersion() {
		return version;
	}
	
	@Override
	public int compareTo(CircuitSimVersion o) {
		if (this.major > o.major) {
			return 1;
		} else if (this.major == o.major) {
			if (this.minor > o.minor) {
				return 1;
			} else if (this.minor == o.minor) {
				if (this.bugfix > o.bugfix) {
					return 1;
				} else if (this.bugfix == o.bugfix) {
					if (this.beta == o.beta) {
						return 0;
					} else if (o.beta) {
						return 1;
					}
				}
			}
		}
		
		return -1;
	}
}

File: src/main/kotlin/com/ra4king/circuitsim/gui/CircuitSimVersion.kt
package com.ra4king.circuitsim.gui

import java.util.regex.Matcher
import java.util.regex.Pattern

class CircuitSimVersion(@JvmField val version: String) : Comparable<CircuitSimVersion?> {
    private val major: Int
    private val minor: Int
    private val bugfix: Int
    private val beta: Boolean

    init {

        val matcher: Matcher = VERSION_PATTERN.matcher(version)

        require(matcher.find()) { "Invalid version string: $version" }

        major = matcher.group(1).toInt()
        minor = matcher.group(2).toInt()
        bugfix = matcher.group(3).toInt()
        beta = !matcher.group(4).isEmpty()
    }

    override fun compareTo(other: CircuitSimVersion?) =
        other?.let {
            (major.compareTo(it.major) * 8 +
                minor.compareTo(it.minor) * 4 +
                bugfix.compareTo(it.bugfix) * 2 +
                it.beta.compareTo(beta)).compareTo(0)
        } ?: 0

    companion object {
        private val VERSION_PATTERN: Pattern = Pattern.compile("(\\d+)\\.(\\d+)\\.(\\d+)(b?)")

        @JvmField
        val VERSION = CircuitSimVersion("1.11.0-CE")
    }
}
